src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -158,9 +189,30 @@ public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
                     // delay build by setting pfDelay to non-zero
                     pfDelay = 1;
                 }
+                // check if no build restore is running, then start a new one
+                else if (!_issRestoreRunning)
+                {
+                    // disable running more build restore
+                    _issRestoreRunning = true;
+
+                    // run build restore
+                    OnBuildRestore();
+
+                    // delay build until restore is running
+                    pfDelay = 1;
```

Comment:
This introduces unnecessary delay when `OnBuildRestore` no-ops. For instance, when performing Clean only, or `ShouldRestoreOnBuild` returns `false`.